src/robot/running/keywordrunner.py

```diff
@@ -206,16 +259,81 @@ def _to_number_with_arithmetics(self, item):
-        result = KeywordResult(kwname=name,
-                               type=data.FOR_ITEM_TYPE)
-        for var, value in zip(data.variables, values):
-            self._context.variables[var] = value
-        runner = KeywordRunner(self._context, self._templated)
-        with StatusReporter(self._context, result):
-            runner.run_keywords(data.keywords)
+
+class ForInZipRunner(ForInRunner):
+    def __init__(self, context, templated=False):
+        super(ForInZipRunner, self).__init__(context, templated)
+
+    def _flavor_name(self):
+        return 'IN ZIP'
+
+    def _replace_variables(self, data):
+        values = super(ForInZipRunner, self)._replace_variables(data)
+        if len(data.variables) == len(data.values):
+            return values
+        raise DataError('FOR IN ZIP expects an equal number of variables and'
+                + ' iterables. Got %d variable%s and %d iterable%s.'
+                % (len(data.variables), s(data.variables),
+                    len(data.values), s(data.values)))
+
+    def _transform_items(self, items):
+        answer = list()
+        for item in items:
+            if not is_list_like(item):
+                # Usually one would expect validation to happen in _validate(),
+                # but this particular validation relies on information that is
+                # not available to _validate().
+                raise DataError("FOR IN ZIP items must all be List-like, got %s."
+                    % type_name(item))
+        for zipped_item in zip(*[list(item) for item in items]):
+            answer.extend(zipped_item)
+        return answer
+
+
+class ForInEnumerateRunner(ForInRunner):
+    def __init__(self, context, templated=False):
+        super(ForInEnumerateRunner, self).__init__(context, templated)
+
+    def _flavor_name(self):
+        return 'IN ENUMERATE'
+
+    def _original_values_per_iteration(self, values):
+        if len(values) < 2:
+            raise DataError('FOR IN ENUMERATE expected 2 or more loop variables, got %d.'
+                    % len(values))
+        return len(values) - 1
+
+    def _get_values_for_one_round(self, data):
+        for idx, values in enumerate(super(ForInEnumerateRunner, self)
+                ._get_values_for_one_round(data)):
+            yield [idx] + values
+
+    def _replace_variables_error_message(self, data, values):
+        return ('Number of FOR IN ENUMERATE loop values (excluding the counter)'
+                + ' should be multiple of its variables.'
+                + ' Got %d non-counting variable%s but %d value%s.'
+                % (len(data.variables), s(data.variables),
+                    len(values), s(values)))
+
+
+class InvalidForRunner(ForInRunner):
+    """
+    Used to send an error from ForRunner() if it sees an unexpected error.
+
+    We can't simply throw a DataError from ForRunner() because that happens
+    outside the "with StatusReporter(...)" blocks.
+
+    """
```

Comment:
The first line should follow the leading quotes like `"""Used to ...` as recommended in [PEP-257](https://www.python.org/dev/peps/pep-0257/). The unnecessary empty line should also be removed.